### PR TITLE
v2.0: massively reduce library size, remove features

### PR DESCRIPTION
**Pull request by paulmillr**
- getPublicKey(), sign(), verify(), getSharedSecret(), Signature.recoverPublicKey()
- Improved security
- Errors are sometimes throws with empty messages and longer stack traces
- No Schnorr: just ECDSA
- No DER: just compact signature representation
- No micro-optimizations, but still almost as fast
- No support for rare JS environments that can't use bigint literals, exponentiation operator

ed25519: https://github.com/paulmillr/noble-ed25519/pull/76

### CODE REVIEW

**Comment by paulmillr** [comment]
Iteration 7:

```
getPublicKey(utils.randomPrivateKey()) x 5,006 ops/sec @ 199μs/op
sign x 4,062 ops/sec @ 246μs/op
verify x 476 ops/sec @ 2ms/op
getSharedSecret x 401 ops/sec @ 2ms/op
Point.fromHex (decompression) x 7,665 ops/sec @ 130μs/op
```

**Comment by MicahZoltu** [comment]
You could save some more space by dropping semicolons.  😆 

On a more serious note, it would be nice if you took the opportunity in this version bump to move the source code to a `source` folder and the build output to an `output` folder.  This makes it easier for third parties and tools to extract just the build output or just the source from the project without needing to do anything special.  It also makes it easier to `clean` by just deleting the output folder.

**Comment by paulmillr** [comment]
index.js is not build output. Not in deno. In deno, index.ts is the single file which is utilized.

**Comment by MicahZoltu** [comment]
I was thinking something like this?  I believe for the output you use `lib` which is a reasonable alternative to `output` if that is your preference, so the primary change here would be moving `index.ts` from root to `source/index.ts`.
```
root
|--source/
|----index.ts
|--output/
|----index.js
|--mod.ts
```
```json
{
	"main": "./output/index.js"
	"exports": {
		".": {
		"types": "./lib/index.d.ts",
		"import": "./lib/index.js",
		"default": "./lib/index.js"
	}
}
```

**Comment by paulmillr** [comment]
I will likely drop common.js and just place it all top-level.

**Comment by MicahZoltu** [comment]
I'm a big fan of dropping CJS!  I still think source code and build output should still end up in a folder though as that makes it easier to work with.  Especially since build output likely includes several files (d.ts, .d.ts.map, js.map, .js) and being able to move them around in bulk without needing fancy glob patterns is nice for custom build tooling.

**Comment by paulmillr** [comment]
Again, the point is that `index.js` is not build output in Deno. In deno, there is no build output.

**Comment by MicahZoltu** [comment]
> Again, the point is that `index.js` is not build output in Deno. In deno, there is no build output.

I understand that, but I don't understand what this has to do with the proposal I'm making?  Why does moving the build output to an output folder and the original source files into a source folder affect Deno?

Is the problem here that Deno can *only* run `<root>/index.ts` and there is no way to tell it to look in `source/index.ts`?  If this is the case could we not just have a `<root>/index.ts` which simply contains `export * from './source/index.ts'`?

**Comment by paulmillr** [comment]
Tests are now using `micro-should` and are shared between `noble/curves`.

File: index.ts [hunk 1]
```diff
@@ -220,7 +220,7 @@ const bits2int_modN = (bytes: Uint8Array): bigint => { // int2octets can't be us
 const i2o = (num: bigint): Bytes => n2b(num);           // int to octets
 declare const globalThis: Record<string, any> | undefined;    // Typescript global symbol available in
```
Comment (MicahZoltu) [inline index.ts:221]: This is dangerous from a type checking standpoint.  It is saying that any undefined variable is available anywhere and should not be type checked at all.

A more type safe alternative would be:
```ts
declare module globalThis {
  const crypto: {
    getRandomValues<T extends Uint8Array>(array: T): T;
    subtle: {
      digest(algorithm: 'SHA-256', data: Uint8Array): Promise<ArrayBuffer>;
      importKey(
        format: 'raw',
        keyData: Uint8Array,
        algorithm: { name: 'HMAC'; hash: { name: 'SHA-256' } },
        extractable: boolean,
        keyUsages: 'sign'[]
      ): Promise<CryptoKey>;
      sign(algorithm: 'HMAC', key: CryptoKey, data: Uint8Array): Promise<ArrayBuffer>;
    };
  };
}
```

File: index.ts [hunk 2]
```diff
@@ -220,7 +220,7 @@ const bits2int_modN = (bytes: Uint8Array): bigint => { // int2octets can't be us
...
 const cr: { node?: any; web?: any } = {     // browsers only. Ensure no dependence on @types/dom
-  node: typeof require === 'function' && require('crypto'), // node.js require('crypto')
+  node: typeof require === 'function' && require('node:crypto'), // node.js require('crypto')
```
Comment (MicahZoltu) [inline index.ts:223]: I'm surprised TypeScript isn't complaining about referencing `require` here without first defining it.  You should be getting an error `Cannot find name 'require'.`  My guess is that you have `@types/node` installed which is incorrectly allowing you to use node types in this project.  Consider removing that dependency and adding the following to make this code not error:
```ts
declare const require: unknown
```

**Comment by Akamig** [comment]
Erm.... no offense and just askin', was there any specific reason about dropping DER format support?
We were using that well at dealing with frustrating AWS-KMS provide key in ASN.1 format.
(of course it's their and our problem to deal with not yours, but well.)

**Comment by paulmillr** [comment]
@Akamig it takes too much code and it's only used in specific situations. You're free to use noble-curves which has DER and is a proper evolution of the current library. goals of new noble-secp are being 1) safe 2) as small as possible 3) fast

**Comment by Akamig** [comment]
@paulmillr Oh, I didn't expected answer this fast. thanks for your quick, kind, understandable response! 🥰

Y E A H    G O T T A    G O    F A S T

**Pull request merged**